lib/src/main/java/graphql/nadel/engine/transform/NadelServiceTypeFilterTransform.kt

```diff
@@ -90,22 +90,27 @@ class NadelServiceTypeFilterTransform : NadelTransform<State> {
 
         val typeNamesOwnedByService = executionBlueprint.getOverAllTypeNamesForService(service)
 
-        val fieldObjectTypeNamesOwnedByService = overallField.objectTypeNames
-            .filter { objectTypeName ->
-                isTypeOwnedByService(
-                    objectTypeName,
-                    executionContext,
-                    service,
-                    executionBlueprint,
-                )
+        // Add underlying type names as well to handle combination of hydration and renames.
+        // Transforms are applied to hydration fields as well, and those fields always reference
+        // elements from the underlying schema
+        val underlyingTypeNamesOwnedByService = executionBlueprint.getUnderlyingTypeNamesForService(service)
+
+        // Assume for most cases there aren't foreign types so there is no point filtering to a new List
+        val noForeignTypes = overallField.objectTypeNames
+            .all { objectTypeName ->
+                objectTypeName in typeNamesOwnedByService
+                    || objectTypeName in underlyingTypeNamesOwnedByService
             }
```

Comment:
This first checks if we have any foreign types before running `.filter` to actually transform it.

Most of the time we should only have types for the service, so we shouldn't need to invoke `.filter` and create a new list with everything copied over.